--log-http consumes the Request Body, making it invalid

<!-- If you need to report a security issue with Knative, send an email to knative-security@googlegroups.com. -->
## In what area(s)?

<!-- Remove the '> ' to select -->

<!--
Classifications:
> /kind good-first-issue
> /kind doc
> /kind cleanup
-->

## What version of Knative Client?

Version:      v20190819-local-dccaf36                                                                                                                                                                      
Build Date:   2019-08-19 22:55:23                                                                                                                                                                          
Git Revision: dccaf36                                                                                                                                                                                      
Support:                                                                                                                                                                                                   
- Serving: v0.8.0  v0.7.1
- API(s):  v1alpha1

## Expected Behavior

I expect that adding the --log-http flag to a command makes no difference to the request the server sees.

## Actual Behavior

Adding the --log-http flag will send empty bodies in all requests (with incorrect content-length headers to boot).

## Expected Cause

https://github.com/knative/client/blob/dccaf366118159b327f9125324c07c5c753b7d21/pkg/util/logging_http_transport.go#L51
Copies the request. Importantly, Request's Body is an interface and therefore a pass by reference. Body can only be read once, so

https://github.com/knative/client/blob/dccaf366118159b327f9125324c07c5c753b7d21/pkg/util/logging_http_transport.go#L65

drains the body for _both_ r and reqCopy. It then sets a copy for reqCopy.


## Steps to Reproduce the Problem

<!-- How can a maintainer reproduce this issue (be detailed) -->


Logs:
```
 ./kn service create --log-http foo --image foo.bar 
===== REQUEST =====
GET /apis/serving.knative.dev/v1alpha1/namespaces/default/services/foo HTTP/1.1
Host: 35.222.62.54
User-Agent: kn/v0.0.0 (linux/amd64) kubernetes/$Format
Accept: application/json, */*
Authorization: *******************************************************************************************************************************************                                                
Accept-Encoding: gzip


===== RESPONSE =====
HTTP/2.0 404 Not Found
Content-Length: 232
Audit-Id: a032637e-a483-4232-abfd-2ed69beae461
Content-Type: application/json
Date: Mon, 19 Aug 2019 22:56:35 GMT

{"kind":"Status","apiVersion":"v1","metadata":{},"status":"Failure","message":"services.serving.knative.dev \"foo\" not found","reason":"NotFound","details":{"name":"foo","group":"serving.knative.dev","kind":"services"},"code":404}

 * * * * * *
===== REQUEST =====
POST /apis/serving.knative.dev/v1alpha1/namespaces/default/services HTTP/1.1
Host: 35.222.62.54
User-Agent: kn/v0.0.0 (linux/amd64) kubernetes/$Format
Content-Length: 298
Accept: application/json, */*
Authorization: *******************************************************************************************************************************************                                                
Content-Type: application/json
Accept-Encoding: gzip

{"kind":"Service","apiVersion":"serving.knative.dev/v1alpha1","metadata":{"name":"foo","namespace":"default","creationTimestamp":null},"spec":{"template":{"metadata":{"name":"foo-wtltm-1","creationTimestamp":null},"spec":{"containers":[{"name":"","image":"foo.bar","resources":{}}]}}},"status":{}}

===== RESPONSE =====
HTTP/2.0 500 Internal Server Error
Content-Length: 156
Audit-Id: ea98f490-abaf-47f3-a445-150a9c2d120b
Content-Type: application/json
Date: Mon, 19 Aug 2019 22:56:35 GMT

{"kind":"Status","apiVersion":"v1","metadata":{},"status":"Failure","message":"request declared a Content-Length of 298 but only wrote 0 bytes","code":500}                                               

 * * * * * *
```

This was followed by an unrelated panic:
```
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x48 pc=0x116bb94]

goroutine 1 [running]:
knative.dev/client/pkg/errors.isCRDError(0x7f748f7b1208, 0xc000264120, 0xc000264120)
        /usr/local/google/home/harwayne/go-mod/github.com/knative/client/pkg/errors/factory.go:25 +0x74                                                                                                   
knative.dev/client/pkg/errors.GetError(0x15986c0, 0xc000264120, 0xc0003e2000, 0x15986c0)
        /usr/local/google/home/harwayne/go-mod/github.com/knative/client/pkg/errors/factory.go:44 +0x8f                                                                                                   
knative.dev/client/pkg/serving/v1alpha1.(*knClient).CreateService(0xc000310260, 0xc0001eae00, 0x915503, 0xc0002a8280)                                                                                     
        /usr/local/google/home/harwayne/go-mod/github.com/knative/client/pkg/serving/v1alpha1/client.go:174 +0x84                                                                                         
knative.dev/client/pkg/kn/commands/service.createService(0x15e6800, 0xc000310260, 0xc0001eae00, 0x13dd649, 0x7, 0x1599000, 0xc000010018, 0x0, 0x0)                                                        
        /usr/local/google/home/harwayne/go-mod/github.com/knative/client/pkg/kn/commands/service/service_create.go:136 +0x4b                                                                              
knative.dev/client/pkg/kn/commands/service.NewServiceCreateCommand.func1(0xc0002a8c80, 0xc0002a2ac0, 0x1, 0x4, 0x0, 0x0)                                                                                  
        /usr/local/google/home/harwayne/go-mod/github.com/knative/client/pkg/kn/commands/service/service_create.go:100 +0x6e2                                                                             
github.com/spf13/cobra.(*Command).execute(0xc0002a8c80, 0xc0002a2a80, 0x4, 0x4, 0xc0002a8c80, 0xc0002a2a80)                                                                                               
        /usr/local/google/home/harwayne/go-mod/github.com/knative/client/vendor/github.com/spf13/cobra/command.go:762 +0x465                                                                              
github.com/spf13/cobra.(*Command).ExecuteC(0xc0002a8280, 0x15bc74fbd460474b, 0x206aa20, 0x51cddf)
        /usr/local/google/home/harwayne/go-mod/github.com/knative/client/vendor/github.com/spf13/cobra/command.go:852 +0x2ec                                                                              
github.com/spf13/cobra.(*Command).Execute(...)
        /usr/local/google/home/harwayne/go-mod/github.com/knative/client/vendor/github.com/spf13/cobra/command.go:800                                                                                     
main.main()
        /usr/local/google/home/harwayne/go-mod/github.com/knative/client/cmd/kn/main.go:36 +0xb6
```

--- a/pkg/util/logging_http_transport.go
+++ b/pkg/util/logging_http_transport.go
@@ -21,6 +21,17 @@ import (
 	"net/http/httputil"
 	"os"
 	"strings"
+
+	"k8s.io/apimachinery/pkg/util/sets"
+)
+
+var (
+	// sensitiveRequestHeaders are headers that will be redacted when logging requests.
+	sensitiveRequestHeaders = sets.NewString(
+		"Authorization",
+		"WWW-Authenticate",
+		"Cookie",
+		"Proxy-Authorization")
 )
 
 type LoggingHttpTransport struct {
@@ -36,35 +47,34 @@ func NewLoggingTransportWithStream(transport http.RoundTripper, s io.Writer) htt
 	return &LoggingHttpTransport{transport, s}
 }
 
-var SENSITIVE_HEADERS = map[string]bool{
-	"Authorization":       true,
-	"WWW-Authenticate":    true,
-	"Cookie":              true,
-	"Proxy-Authorization": true,
-}
-
 func (t *LoggingHttpTransport) RoundTrip(r *http.Request) (*http.Response, error) {
 	stream := t.stream
 	if stream == nil {
 		stream = os.Stderr
 	}
-	reqCopy := *r
-	reqCopy.Header = make(http.Header, len(r.Header))
+	redacted := http.Header{}
 	for k, v := range r.Header {
-		sensitive := SENSITIVE_HEADERS[k]
-		if sensitive {
+		if sensitiveRequestHeaders.Has(k) {
+			redacted[k] = v
 			l := 0
 			for _, h := range v {
 				l += len(h)
 			}
-			reqCopy.Header.Set(k, strings.Repeat("*", l))
-		} else {
-			reqCopy.Header[k] = v
+			r.Header.Set(k, strings.Repeat("*", l))
 		}
 	}
-	reqBytes, _ := httputil.DumpRequestOut(&reqCopy, true)
+	reqBytes, err := httputil.DumpRequestOut(r, true)
+	if err != nil {
+		fmt.Fprintln(stream, "error dumping request:", err)
+		return nil, fmt.Errorf("dumping request: %v", err)
+	}
 	fmt.Fprintln(stream, "===== REQUEST =====")
 	fmt.Fprintln(stream, string(reqBytes))
+
+	for k, v := range redacted {
+		r.Header[k] = v
+	}
+
 	resp, err := t.transport.RoundTrip(r)
 	if err != nil {
 		fmt.Fprintln(stream, "===== ERROR =====")

--- a/pkg/util/logging_http_transport_test.go
+++ b/pkg/util/logging_http_transport_test.go
@@ -55,15 +55,26 @@ func TestWritesRequestResponse(t *testing.T) {
 	out := &bytes.Buffer{}
 	dt := &dummyTransport{}
 	transport := NewLoggingTransportWithStream(dt, out)
+
 	body := "{this: is, the: body, of: [the, request]}"
 	req, _ := http.NewRequest("POST", "http://example.com", strings.NewReader(body))
+	nonRedacted := "this string will be logged"
+	redacted := "this string will be redacted"
+	req.Header.Add("non-redacted", nonRedacted)
+	req.Header.Add(sensitiveRequestHeaders.List()[0], redacted)
+
 	_, e := transport.RoundTrip(req)
 	assert.NilError(t, e)
 	s := out.String()
 	assert.Assert(t, strings.Contains(s, "REQUEST"))
 	assert.Assert(t, strings.Contains(s, "RESPONSE"))
 	assert.Assert(t, strings.Contains(s, body))
+	assert.Assert(t, strings.Contains(s, nonRedacted))
+	assert.Assert(t, !strings.Contains(s, redacted))
+
 	assert.Assert(t, strings.Contains(dt.requestDump, body))
+	assert.Assert(t, strings.Contains(dt.requestDump, nonRedacted))
+	assert.Assert(t, strings.Contains(dt.requestDump, redacted))
 }
 
 func TestElideAuthorizationHeader(t *testing.T) {

--- a/pkg/util/logging_http_transport_test.go
+++ b/pkg/util/logging_http_transport_test.go
@@ -17,17 +17,26 @@ package util
 import (
 	"bytes"
 	"errors"
+	"fmt"
 	"io/ioutil"
 	"net/http"
+	"net/http/httputil"
 	"strings"
 	"testing"
 
 	"gotest.tools/assert"
 )
 
-type dummyTransport struct{}
+type dummyTransport struct {
+	requestDump string
+}
 
 func (d *dummyTransport) RoundTrip(r *http.Request) (*http.Response, error) {
+	dump, err := httputil.DumpRequest(r, true)
+	if err != nil {
+		return nil, fmt.Errorf("dumping request: %v", err)
+	}
+	d.requestDump = string(dump)
 	return &http.Response{
 		Status:     "200 OK",
 		StatusCode: 200,
@@ -44,13 +53,17 @@ func (d *errorTransport) RoundTrip(r *http.Request) (*http.Response, error) {
 
 func TestWritesRequestResponse(t *testing.T) {
 	out := &bytes.Buffer{}
-	transport := NewLoggingTransportWithStream(&dummyTransport{}, out)
-	req, _ := http.NewRequest("GET", "http://example.com", nil)
+	dt := &dummyTransport{}
+	transport := NewLoggingTransportWithStream(dt, out)
+	body := "{this: is, the: body, of: [the, request]}"
+	req, _ := http.NewRequest("POST", "http://example.com", strings.NewReader(body))
 	_, e := transport.RoundTrip(req)
 	assert.NilError(t, e)
 	s := out.String()
 	assert.Assert(t, strings.Contains(s, "REQUEST"))
 	assert.Assert(t, strings.Contains(s, "RESPONSE"))
+	assert.Assert(t, strings.Contains(s, body))
+	assert.Assert(t, strings.Contains(dt.requestDump, body))
 }
 
 func TestElideAuthorizationHeader(t *testing.T) {

